src/Pos/Update/Poll/Pure.hs

```diff
@@ -0,0 +1,88 @@
+-- | Pure Poll
+
+module Pos.Update.Poll.Pure
+       ( PurePoll (..)
+       ) where
+
+import           Universum
+
+import           Control.Lens               (at, to, (%=), (.=))
+import qualified Data.HashMap.Strict        as HM
+import           System.Wlog                (CanLog, HasLoggerName, NamedPureLogger,
+                                             logWarning)
+
+import           Pos.Crypto                 (hash)
+import           Pos.Types                  (SoftwareVersion (..))
+import           Pos.Update.Poll.Class      (MonadPoll (..), MonadPollRead (..))
+import qualified Pos.Update.Poll.PollState  as Poll
+import           Pos.Update.Poll.Types      (BlockVersionState (..),
+                                             DecidedProposalState (..),
+                                             UndecidedProposalState (..),
+                                             cpsSoftwareVersion, propStateToEither)
+
+newtype PurePoll a = PurePoll
+    { getPurePoll :: NamedPureLogger (State Poll.PollState) a
+    } deriving (Functor, Applicative, Monad, CanLog, HasLoggerName)
+
+instance MonadPollRead PurePoll where
+    getBVState bv = PurePoll $ use $ Poll.psBlockVersions . at bv
+    getProposedBVs = PurePoll $ use $ Poll.psBlockVersions . to HM.keys
+    getEpochProposers = PurePoll $ use $ Poll.psEpochProposers
+    getConfirmedBVStates = PurePoll $ use $ Poll.psBlockVersions . to HM.toList
+    getAdoptedBVFull = PurePoll $ use $ Poll.psAdoptedBV
+    getLastConfirmedSV an = PurePoll $ use $ Poll.psConfirmedBVs . at an
+    getProposal ui = PurePoll $ use $ Poll.psActiveProposals . at ui
+    getProposalsByApp an = PurePoll $ do
+        delActiveProposalsIndices <- use Poll.psDelActivePropsIdx
+        activeProposals           <- use Poll.psActiveProposals
+        pure $ propGetByApp delActiveProposalsIndices activeProposals
+      where propGetByApp appHashmap upIdHashmap = fromMaybe [] $
+                do hashset <- HM.lookup an appHashmap
+                   let uidList = toList hashset
+                       propStateList = fmap (flip HM.lookup upIdHashmap) uidList
```

Comment:
I would also log warning if some `lookup` fails here. It means that `psDelActivePropsIdx` contains `Upid` which is not in active proposals. It sounds like invariant violation.